core/src/main/java/org/openedx/core/domain/interactor/IAPInteractor.kt

```diff
@@ -1,23 +1,73 @@
 package org.openedx.core.domain.interactor
 
+import android.content.Context
 import androidx.fragment.app.FragmentActivity
 import com.android.billingclient.api.BillingClient.BillingResponseCode
 import com.android.billingclient.api.ProductDetails
 import com.android.billingclient.api.Purchase
 import org.openedx.core.ApiConstants
+import org.openedx.core.R
+import org.openedx.core.config.Config
 import org.openedx.core.data.repository.iap.IAPRepository
+import org.openedx.core.data.storage.CorePreferences
 import org.openedx.core.domain.model.iap.ProductInfo
 import org.openedx.core.exception.iap.IAPException
 import org.openedx.core.extension.decodeToLong
 import org.openedx.core.module.billing.BillingProcessor
 import org.openedx.core.module.billing.getCourseSku
 import org.openedx.core.module.billing.getPriceAmount
+import org.openedx.core.presentation.IAPAnalytics
+import org.openedx.core.presentation.IAPAnalyticsEvent
+import org.openedx.core.presentation.IAPAnalyticsKeys
+import org.openedx.core.presentation.IAPAnalyticsScreen
+import org.openedx.core.presentation.global.AppData
+import org.openedx.core.presentation.iap.IAPAction
+import org.openedx.core.presentation.iap.IAPFlow
 import org.openedx.core.presentation.iap.IAPRequestType
+import org.openedx.core.utils.EmailUtil
 
 class IAPInteractor(
+    private val appData: AppData,
     private val billingProcessor: BillingProcessor,
+    private val config: Config,
     private val repository: IAPRepository,
+    private val preferencesManager: CorePreferences,
+    private val iapAnalytics: IAPAnalytics,
 ) {
+    private val iapConfig
+        get() = preferencesManager.appConfig.iapConfig
+    private val isIAPEnabled
+        get() = iapConfig.isEnabled &&
+            iapConfig.disableVersions.contains(appData.versionName).not()
+
+    fun logIAPCancelEvent(screen: IAPAnalyticsScreen) {
+        logIAPEvent(
+            IAPAnalyticsEvent.IAP_ERROR_ALERT_ACTION, buildMap {
+                put(IAPAnalyticsKeys.ERROR_ALERT_TYPE.key, IAPAction.ACTION_UNFULFILLED.action)
+                put(IAPAnalyticsKeys.ERROR_ACTION.key, IAPAction.ACTION_CLOSE.action)
+            }.toMutableMap(),
+            screen = screen,
+        )
+    }
+
+    fun showFeedbackScreen(context: Context, message: String, screen: IAPAnalyticsScreen) {
```

Comment:
Although it facilitates usage, I'm not entirely comfortable with having context this deep in the domain layer. 🤐

Can we avoid it?

Reply:
Plus please move the `logIAPCancelEvent` and `showFeedbackScreen` to the bottom of the file before `logIAPEvent` method.

Reply:
> Plus please move the `logIAPCancelEvent` and `showFeedbackScreen` to the bottom of the file before `logIAPEvent` method.

There is logic with sorting methods, public top, private bottom

Reply:
> Although it facilitates usage, I'm not entirely comfortable with having context and analytics calling this deep in the domain layer. 🤐
> 
> Can we avoid it?

I agree with you about putting context inside domain layer, but I think here we can use it, because we are using it only as argument, also it is needed only to execute Util function, which has context as an argument

Reply:
Apart from this, having Analytics in the domain layer is a bit different from what we have done so far.

Reply:
Removed analytics from interactor